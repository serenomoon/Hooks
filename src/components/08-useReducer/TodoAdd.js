import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {


    //uso de custom hook
    const [{description}, handleInputChange, reset] = useForm({
        description: ''
      });


    const handleSubmit = (e) =>{
        e.preventDefault();

        if(description.trim().length <= 1){
            return;
        };

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };

        //toma el objeto newTodo y lo agrega a otro, que se usara para 
        //enviar al useReducer, aclarando que es para agregar el nuevo objeto ('add').
        //lo envia al reducer.
        handleAddTodo( newTodo );

        //resetea el valor del input usando el custom hook useForm.
        reset();

    }

  return (
    <>
        <h4>Agregar ToDo</h4>
        <hr />

        <form onSubmit={ handleSubmit }>
            <input 
                type='text'
                name='description'
                className='form-control'
                placeholder='Tengo que...'
                autoComplete='off'
                value={ description }
                onChange={ handleInputChange }
            />

            <button
                type='submit'
                className='btn btn-outline-primary mt-1 btn-block'
                >
                Agregar
            </button>
        </form>
    </>
  )
}
