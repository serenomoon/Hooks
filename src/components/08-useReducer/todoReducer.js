
//toma el estado del objeto en este caso y la accion
export const todoReducer = ( state= [] , action ) => {
  
    switch (action.type) {
        case 'add':
            return [ ...state, action.payload ];

        //filtra para que conserve todos los objetos que tengan un id diferente al seleccionado
        case 'delete':
            return state.filter( todo => todo.id !== action.payload );


        //codigo mejorado del toggle
        case 'toggle':
            return state.map( todo => 
                ( todo.id === action.payload )
                    ? {...todo, done: !todo.done}
                    : todo    
            );

        case 'toggle-old':
            return state.map( todo => {
                if (todo.id === action.payload) {
                    return {
                        ...todo,
                        done: !todo.done
                    }
                } else {
                    return todo;
                }
            });
    
        default:
            return state;
    }

}
