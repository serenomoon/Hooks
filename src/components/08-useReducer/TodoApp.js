import React, { useEffect, useReducer } from 'react'
import { todoReducer } from './todoReducer';

import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';
import './TodoApp.css';


const init = () => {
  //toma los datos almacenados en el localstorage del navegador
  return JSON.parse(localStorage.getItem('todos')) || [];
  
};


export const TodoApp = () => {

    const [ todos, dispatch ] = useReducer(todoReducer, [], init);


    useEffect(() => {
      //si 'todos' cambia, cambia los datos en el localstorage del navegador (almacena o borra)
      localStorage.setItem('todos', JSON.stringify( todos ));
    }, [todos])
    

    const handleAddTodo = (todoAdd) => {

      dispatch({
        type: 'add',
        payload: todoAdd
      })
    }


    const handleDelete = ( todoId ) => {

      dispatch( {
        type: 'delete',
        payload: todoId
      } );
}


    const handleToggle = ( todoId ) => {

      dispatch({
        type: 'toggle',
        payload: todoId
      })
    }

  return (
    <div>
        <h1>TodoApp ( { todos.length } ) </h1>
        <hr />

        <div className='row'>

          <div className='col-7'>

              <TodoList 
                todos= {todos}
                handleToggle= {handleToggle}
                handleDelete= {handleDelete}
              />

          </div>

          <div className='col-5'>

                <TodoAdd 
                handleAddTodo = { handleAddTodo }
                />

          </div>

        </div>    
    </div>
  )
}
