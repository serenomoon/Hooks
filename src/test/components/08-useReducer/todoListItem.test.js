import { shallow } from "enzyme";
import { TodoListItem } from '../../../components/08-useReducer/TodoListItem';
import { demoTodos } from "../../fixtures/demoTodos";


describe('Test <TodoListItem />', () => {
    
    const handleToggle = jest.fn();
    const handleDelete = jest.fn();

    const wrapper = shallow( 
    < TodoListItem 
        todo = { demoTodos[0] }
        index = { 1 }
        handleToggle = { handleToggle }
        handleDelete = { handleDelete }

    /> );

    test('should show correctly', () => {
        
        expect( wrapper ).toMatchSnapshot();

    });

    test('should call the function handleDelete', () => {
        
        wrapper.find('button').simulate('click');

        expect( handleDelete ).toHaveBeenCalled();
        expect( handleDelete ).toHaveBeenCalledWith(1);

    });

    test('should call the function handleToggle', () => {
        
        wrapper.find('p').simulate('click');

        expect( handleToggle ).toHaveBeenCalledWith( demoTodos[0].id );

    });

    test('should show the text correctly', () => {
        
        expect( wrapper.find('p').text().trim() ).toBe(`2. ${demoTodos[0].desc}`)

    });

    test('should have the class "complete" if TODO.done = true', () => {
        
        const todo = demoTodos[0];
        todo.done = true;

        const wrapper = shallow( 
            < TodoListItem 
                todo = { todo }
            /> );

        expect(wrapper.find('p').hasClass("complete")).toBe(true);

    });
});