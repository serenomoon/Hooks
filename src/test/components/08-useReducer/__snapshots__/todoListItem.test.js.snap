// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <TodoListItem /> should show correctly 1`] = `
<li
  className="list-group-item"
  key="1"
>
  <p
    className=" false "
    onClick={[Function]}
  >
    2
    . 
    Lear React
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Borrar
  </button>
</li>
`;
